presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeJoinsOverMapColumns.java

```diff
@@ -0,0 +1,455 @@
+                nullComparisonBuilder.add(nullComp);
+                coalesceComparisonBuilder.add(coalesceComp);
+            }
+
+            ProjectNode projectNode = new ProjectNode(planNodeIdAllocator.getNextId(), joinNode, assignments.build());
+            List<VariableReferenceExpression> resultOutputCols = Stream.concat(planNode.getOutputVariables().stream(), projectNode.getOutputVariables().stream()).collect(toImmutableList());
+
+            List<RowExpression> joinCriteria = Stream.concat(nullComparisonBuilder.build().stream(), coalesceComparisonBuilder.build().stream()).collect(toImmutableList());
+            JoinNode newJoinNode = new JoinNode(
+                    joinNode.getSourceLocation(),
+                    planNodeIdAllocator.getNextId(),
+                    JoinNode.Type.LEFT,
+                    planNode,
+                    projectNode,
+                    ImmutableList.of(),
+                    resultOutputCols,
+                    Optional.of(LogicalRowExpressions.and(joinCriteria)),
+                    joinNode.getLeftHashVariable(),
+                    joinNode.getRightHashVariable(),
+                    joinNode.getDistributionType(),
+                    joinNode.getDynamicFilters());
+
+            return newJoinNode;
+        }
+
+        boolean validateProjectAssignments(ProjectNode projectNode)
+        {
+            Assignments assignments = projectNode.getAssignments();
+            PlanNode input = projectNode.getSource();
+            ImmutableSet<VariableReferenceExpression> inputColsSet = ImmutableSet.copyOf(input.getOutputVariables());
+
+            for (Map.Entry<VariableReferenceExpression, RowExpression> assignment : assignments.entrySet()) {
+                RowExpression expr = assignment.getValue();
+                if (!inputColsSet.containsAll(VariablesExtractor.extractUnique(expr))) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private RowExpression coalesceToZero(RowExpression var)
+        {
+            RowExpression zero = constant(0L, BIGINT);
+            return coalesce(ImmutableList.of(var, zero));
+        }
+
+        private Set<VariableReferenceExpression> extractJoinKeys(Optional<RowExpression> filter, List<JoinNode.EquiJoinClause> criteria)
+        {
+            ImmutableSet.Builder<VariableReferenceExpression> builder = ImmutableSet.builder();
+
+            criteria.forEach((v) -> {
+                builder.add(v.getLeft());
+                builder.add(v.getRight());
+            });
+
+            if (filter.isPresent()) {
+                builder.addAll(VariablesExtractor.extractAll(filter.get()));
+            }
+
+            return builder.build();
+        }
+
+        private boolean supportedJoinKeyTypes(Set<VariableReferenceExpression> joinKeys)
+        {
+            return joinKeys.stream().allMatch(key -> TypeUtils.isNumericType(key.getType()));
```

Comment:
I am wondering why this is limited to NumericType only?

Reply:
mostly because of the coalesce handling for NULL values, can extend it to handle other simple types like strings but not sure if it will be useful for production queries

Reply:
Lets handle other types as well, if we leave it now we will probably never come back to it. :-)